crates/bevy_ecs/src/component.rs

```diff
@@ -1983,35 +1989,69 @@ impl RequiredComponents {
         constructor: fn() -> C,
         inheritance_depth: u16,
     ) {
-        let erased: RequiredComponentConstructor = RequiredComponentConstructor(Arc::new(
-            move |table,
-                  sparse_sets,
-                  change_tick,
-                  table_row,
-                  entity,
-                  #[cfg(feature = "track_change_detection")] caller| {
-                OwningPtr::make(constructor(), |ptr| {
-                    // SAFETY: This will only be called in the context of `BundleInfo::write_components`, which will
-                    // pass in a valid table_row and entity requiring a C constructor
-                    // C::STORAGE_TYPE is the storage type associated with `component_id` / `C`
-                    // `ptr` points to valid `C` data, which matches the type associated with `component_id`
-                    unsafe {
-                        BundleInfo::initialize_required_component(
-                            table,
-                            sparse_sets,
-                            change_tick,
-                            table_row,
-                            entity,
-                            component_id,
-                            C::STORAGE_TYPE,
-                            ptr,
-                            #[cfg(feature = "track_change_detection")]
-                            caller,
-                        );
-                    }
-                });
-            },
-        ));
+        let erased: RequiredComponentConstructor = RequiredComponentConstructor({
+            // `portable-atomic-util` `Arc` is not able to coerce an unsized
+            // type like `std::sync::Arc` can. Creating a `Box` first does the
+            // coercion.
+            //
+            // This would be resolved by https://github.com/rust-lang/rust/issues/123430
```

Comment:
This area is a little frustrating. In short, `std::sync::Arc` has certain privileges that `portable_atomic_util::Arc` doesn't, the key one being unsized coercion. To work around this, we first create a `Box` (which can do the coercion), and then create an `Arc` from that box. This code is a little verbose to handle the 2×2 feature matrix of tracking changes and portable atomics while also ensuring the `Box` workaround is only used when `portable-atomic` is enabled.

Reply:
Might be good to note that `#[derive(CoercePointee)]` is currently in FCP for stabilization, which means it is likely that we'll have it by 1.85 (20th February)

Reply:
Would it be easier to combine the code if the closure inside `RequiredComponentConstructor` took a `MaybeLocation`?  The comments on the alias say "Please use this type sparingly", but I'm not sure what actual cost there is to using it.

Reply:
I _do_ think this should be used in more places. I don't like the verbosity that the `cfg(...)` introduces, and I don't believe that it's more performant (a ZST _should_ be removable from a function argument at compile time, since it has zero size)

Reply:
It'd be good to note that the feature that resolves this, `[derive(CoercePointee)]` is currently in FCP for stabilization, so it will likely hit stable in about 2 months.